File: tests/Explorer.Core.Tests/ViewModels/TabViewModelTests.cs
using NoctusExplorer.Core.Abstractions;
using NoctusExplorer.Core.Models;
using NoctusExplorer.Core.ViewModels;
using NSubstitute;
using FluentAssertions;

namespace NoctusExplorer.Core.Tests.ViewModels;

public class TabViewModelTests
{
    private static (TabViewModel vm, IShellService shell, IFileWatcher watcher) CreateTab(string path = "/home")
    {
        var shell = Substitute.For<IShellService>();
        var watcher = Substitute.For<IFileWatcher>();
        var location = new PathRef(path, isDirectory: true);
        shell.GetDisplayName(Arg.Any<PathRef>()).Returns(ci => ci.Arg<PathRef>().DisplayName);
        shell.EnumerateAsync(Arg.Any<PathRef>(), Arg.Any<CancellationToken>())
            .Returns(Task.FromResult<IReadOnlyList<FileEntry>>([]));

        var vm = new TabViewModel(0, location, shell, watcher);
        return (vm, shell, watcher);
    }

    [Fact]
    public void Constructor_SetsInitialLocation()
    {
        var (vm, _, _) = CreateTab("/home");
        vm.Location.FullPath.Should().Be("/home");
    }

    [Fact]
    public async Task NavigateAsync_UpdatesLocation()
    {
        var (vm, shell, _) = CreateTab();
        var target = new PathRef("/docs", isDirectory: true);
        shell.EnumerateAsync(target, Arg.Any<CancellationToken>())
            .Returns(Task.FromResult<IReadOnlyList<FileEntry>>([]));

        await vm.NavigateAsync(target);
        vm.Location.Should().Be(target);
    }

    [Fact]
    public async Task NavigateAsync_PopulatesEntries()
    {
        var (vm, shell, _) = CreateTab();
        var target = new PathRef("/docs", isDirectory: true);
        var entries = new List<FileEntry>
        {
            new(new PathRef("/docs/file.txt"), "file.txt", ".txt", 100,
                DateTimeOffset.Now, DateTimeOffset.Now, false, false, "Document")
        };
        shell.EnumerateAsync(target, Arg.Any<CancellationToken>())
            .Returns(Task.FromResult<IReadOnlyList<FileEntry>>(entries));

        await vm.NavigateAsync(target);
        vm.Entries.Should().HaveCount(1);
        vm.Entries[0].Name.Should().Be("file.txt");
    }

    [Fact]
    public async Task NavigateAsync_WatchesNewDirectory()
    {
        var (vm, shell, watcher) = CreateTab();
        var target = new PathRef("/docs", isDirectory: true);
        shell.EnumerateAsync(target, Arg.Any<CancellationToken>())
            .Returns(Task.FromResult<IReadOnlyList<FileEntry>>([]));

        await vm.NavigateAsync(target);
        watcher.Received(1).Watch(target);
    }

    [Fact]
    public async Task NavigateAsync_UnwatchesPreviousDirectory()
    {
        var (vm, shell, watcher) = CreateTab("/home");
        var target = new PathRef("/docs", isDirectory: true);
        shell.EnumerateAsync(target, Arg.Any<CancellationToken>())
            .Returns(Task.FromResult<IReadOnlyList<FileEntry>>([]));

        var original = vm.Location;
        await vm.NavigateAsync(target);
        watcher.Received(1).Unwatch(original);
    }

    [Fact]
    public async Task GoBackAsync_NavigatesToPreviousLocation()
    {
        var (vm, shell, _) = CreateTab("/a");
        var b = new PathRef("/b", isDirectory: true);
        shell.EnumerateAsync(Arg.Any<PathRef>(), Arg.Any<CancellationToken>())
            .Returns(Task.FromResult<IReadOnlyList<FileEntry>>([]));

        await vm.NavigateAsync(b);
        await vm.GoBackAsync();
        vm.Location.FullPath.Should().Be("/a");
    }

    [Fact]
    public async Task GoForwardAsync_NavigatesToNextLocation()
    {
        var (vm, shell, _) = CreateTab("/a");
        var b = new PathRef("/b", isDirectory: true);
        shell.EnumerateAsync(Arg.Any<PathRef>(), Arg.Any<CancellationToken>())
            .Returns(Task.FromResult<IReadOnlyList<FileEntry>>([]));

        await vm.NavigateAsync(b);
        await vm.GoBackAsync();
        await vm.GoForwardAsync();
        vm.Location.FullPath.Should().Be("/b");
    }

    [Fact]
    public async Task GoUpAsync_NavigatesToParent()
    {
        var (vm, shell, _) = CreateTab("/home/user/docs");
        shell.EnumerateAsync(Arg.Any<PathRef>(), Arg.Any<CancellationToken>())
            .Returns(Task.FromResult<IReadOnlyList<FileEntry>>([]));

        await vm.GoUpAsync();
        vm.Location.FullPath.Should().Be("/home/user");
    }

    [Fact]
    public void UpdateSelectionSummary_NoSelection_ShowsItemCount()
    {
        var (vm, _, _) = CreateTab();
        vm.Entries.Add(new FileEntry(new PathRef("/a.txt"), "a.txt", ".txt", 100,
            DateTimeOffset.Now, DateTimeOffset.Now, false, false, "Doc"));
        vm.UpdateSelectionSummary();
        vm.SelectionSummary.Should().Be("1 items");
    }

    [Fact]
    public void UpdateSelectionSummary_WithSelection_ShowsCountAndSize()
    {
        var (vm, _, _) = CreateTab();
        var entry = new FileEntry(new PathRef("/a.txt"), "a.txt", ".txt", 2048,
            DateTimeOffset.Now, DateTimeOffset.Now, false, false, "Doc");
        vm.Entries.Add(entry);
        vm.Selection.Add(entry);
        vm.UpdateSelectionSummary();
        vm.SelectionSummary.Should().Be("1 selected (2.0 KB)");
    }

    [Fact]
    public void SetFilter_ActivatesFilter()
    {
        var (vm, _, _) = CreateTab();
        vm.SetFilter("*.txt");
        vm.FilterText.Should().Be("*.txt");
        vm.IsFilterActive.Should().BeTrue();
    }

    [Fact]
    public void ClearFilter_DeactivatesFilter()
    {
        var (vm, _, _) = CreateTab();
        vm.SetFilter("*.txt");
        vm.ClearFilter();
        vm.FilterText.Should().BeEmpty();
        vm.IsFilterActive.Should().BeFalse();
    }

    [Fact]
    public void TogglePreview_FlipsVisibility()
    {
        var (vm, _, _) = CreateTab();
        vm.IsPreviewVisible.Should().BeFalse();
        vm.TogglePreview();
        vm.IsPreviewVisible.Should().BeTrue();
        vm.TogglePreview();
        vm.IsPreviewVisible.Should().BeFalse();
    }
}

File: macos/Tests/NoctusCoreTests/TabViewModelTests.swift
import Testing
import Foundation
@testable import NoctusCore

@Suite struct TabViewModelTests {

    private func createTab(path: String = "/home") -> (vm: TabViewModel, shell: MockShellService, watcher: MockFileWatcher) {
        let shell = MockShellService()
        let watcher = MockFileWatcher()
        let location = PathRef(path, isDirectory: true)
        let vm = TabViewModel(id: 0, location: location, shellService: shell, fileWatcher: watcher)
        return (vm, shell, watcher)
    }

    private func makeEntry(path: String, name: String, ext: String, size: Int64?) -> FileEntry {
        FileEntry(
            path: PathRef(path),
            name: name,
            ext: ext,
            size: size,
            dateModified: Date(),
            dateCreated: Date(),
            isHidden: false,
            kind: "Document"
        )
    }

    @Test func constructorSetsInitialLocation() {
        let (vm, _, _) = createTab(path: "/home")
        #expect(vm.location.fullPath == "/home")
    }

    @Test func navigateUpdatesLocation() async throws {
        let (vm, _, _) = createTab()
        let target = PathRef("/docs", isDirectory: true)
        try await vm.navigate(to: target)
        #expect(vm.location == target)
    }

    @Test func navigatePopulatesEntries() async throws {
        let (vm, shell, _) = createTab()
        let target = PathRef("/docs", isDirectory: true)
        shell.enumerateResult = [makeEntry(path: "/docs/file.txt", name: "file.txt", ext: ".txt", size: 100)]
        try await vm.navigate(to: target)
        #expect(vm.entries.count == 1)
        #expect(vm.entries[0].name == "file.txt")
    }

    @Test func navigateWatchesNewDirectory() async throws {
        let (vm, _, watcher) = createTab()
        let target = PathRef("/docs", isDirectory: true)
        try await vm.navigate(to: target)
        #expect(watcher.watchCalls.contains(target))
    }

    @Test func navigateUnwatchesPreviousDirectory() async throws {
        let (vm, _, watcher) = createTab(path: "/home")
        let original = vm.location
        let target = PathRef("/docs", isDirectory: true)
        try await vm.navigate(to: target)
        #expect(watcher.unwatchCalls.contains(original))
    }

    @Test func goBackNavigatesToPreviousLocation() async throws {
        let (vm, _, _) = createTab(path: "/a")
        let b = PathRef("/b", isDirectory: true)
        try await vm.navigate(to: b)
        try await vm.goBack()
        #expect(vm.location.fullPath == "/a")
    }

    @Test func goForwardNavigatesToNextLocation() async throws {
        let (vm, _, _) = createTab(path: "/a")
        let b = PathRef("/b", isDirectory: true)
        try await vm.navigate(to: b)
        try await vm.goBack()
        try await vm.goForward()
        #expect(vm.location.fullPath == "/b")
    }

    @Test func goUpNavigatesToParent() async throws {
        let (vm, _, _) = createTab(path: "/home/user/docs")
        try await vm.goUp()
        #expect(vm.location.fullPath == "/home/user")
    }

    @Test func updateSelectionSummaryNoSelectionShowsItemCount() {
        let (vm, _, _) = createTab()
        vm.entries.append(makeEntry(path: "/a.txt", name: "a.txt", ext: ".txt", size: 100))
        vm.updateSelectionSummary()
        #expect(vm.selectionSummary == "1 items")
    }

    @Test func updateSelectionSummaryWithSelectionShowsCountAndSize() {
        let (vm, _, _) = createTab()
        let entry = makeEntry(path: "/a.txt", name: "a.txt", ext: ".txt", size: 2048)
        vm.entries.append(entry)
        vm.selection.append(entry)
        vm.updateSelectionSummary()
        #expect(vm.selectionSummary == "1 selected (2.0 KB)")
    }

    @Test func setFilterActivatesFilter() {
        let (vm, _, _) = createTab()
        vm.setFilter("*.txt")
        #expect(vm.filterText == "*.txt")
        #expect(vm.isFilterActive == true)
    }

    @Test func clearFilterDeactivatesFilter() {
        let (vm, _, _) = createTab()
        vm.setFilter("*.txt")
        vm.clearFilter()
        #expect(vm.filterText == "")
        #expect(vm.isFilterActive == false)
    }

    @Test func togglePreviewFlipsVisibility() {
        let (vm, _, _) = createTab()
        #expect(vm.isPreviewVisible == false)
        vm.togglePreview()
        #expect(vm.isPreviewVisible == true)
        vm.togglePreview()
        #expect(vm.isPreviewVisible == false)
    }
}
